### DISCUSSION TITLE
How to inject a middleware correctly having the route info?

### DISCUSSION THREAD

**Discussion by jorge-ribeiro** [ThreeMammals/Ocelot#2118] [Q&A] [answered]
I have an Ocelot web API solution in which requests must be checked by a `CaptchaPayloadMiddleware` to verify if they have the proper custom captcha key/value in a request header, before the request is forwarded to business entities webapis.
When captcha is not present or invalid an error response is sent back. I also have to be able to get the selected downstream route in order to get some [metadata](https://ocelot.readthedocs.io/en/latest/features/metadata.html) from `ocelot.json` file.

The thing is when i add my middleware like:
```C#
app.UseMiddleware<CaptchaPayloadMiddleware>();
app.UseOcelot().Wait();

```

The response is sent back nicely but i cannot access route info, and when i add my middleware like:

```C#
app.UseOcelot().Wait();
app.UseMiddleware<CaptchaPayloadMiddleware>();

```

I have access to route info, but cannot change response as it already started.
It sends the response anyway but with an ugly warning: `ReasonPhrase cannot be set because the response has already started`

**Is this even possible to do with Ocelot?**
If so can you help me here?

Thanks
Regards

**Comment by raman-m** [comment, marked as answer]
Hello Jorge!
Welcome to Ocelot world! 🐯 

> The thing is when I add my middleware like:
> ```C#
> app.UseMiddleware<CaptchaPayloadMiddleware>();
> app.UseOcelot().Wait();
> ```
> The response is sent back nicely but i cannot access route info...

To make route information available in the `HttpContext`, including [Route Metadata](https://ocelot.readthedocs.io/en/latest/features/configuration.html#route-metadata), you need to:
- Convert the `CaptchaPayloadMiddleware` into an anonymous or named `Func<HttpContext, Func<Task>, Task>` delegate.
- Attach the `Func` delegate to an `OcelotPipelineConfiguration` instance within the `UseOcelot` method, utilizing the [Middleware Injection](https://ocelot.readthedocs.io/en/latest/features/middlewareinjection.html) feature.

Please be aware that Ocelot utilizes a unique internal middleware signature, which differs from the classic ASP.NET middleware. For reference, see the [OcelotMiddlewareExtensions](https://github.com/search?q=repo%3AThreeMammals%2FOcelot%20OcelotMiddlewareExtensions&type=code) and [OcelotPipelineExtensions](https://github.com/search?q=repo%3AThreeMammals%2FOcelot+OcelotPipelineExtensions&type=code) classes. The Ocelot kernel, which is built upon the ASP.NET middleware architecture, processes requests. If you insert your own middleware prior to Ocelot's, no information will be available as the request has not yet been processed by the Ocelot kernel. Therefore, it is recommended to use the  [Middleware Injection](https://ocelot.readthedocs.io/en/latest/features/middlewareinjection.html) feature exclusively.

---
> ...and when I add my middleware like:
> ```C#
> app.UseOcelot().Wait();
> app.UseMiddleware<CaptchaPayloadMiddleware>();
> ```
> I have access to route info, but cannot change response as it already started.
> It sends the response anyway but with an ugly warning: `ReasonPhrase cannot be set because the response has already started`

Certainly! Adding your own middleware after the `UseOcelot` call will terminate the request processing by Ocelot kernel, making it impossible to modify the response. To alter the response effectively, utilize the [Delegating Handler](https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html) feature. In particular, you will need to create your own `HttpResponseMessage` object.

I hope this is helpful!
Please reach out if anything is unclear.
